File: go/binary-tree-tilt.go
//
//
package leetcode

var tilt int = 0

func findTilt(root *TreeNode) int {
	postorder(root)
	return tilt
}

func postorder(node *TreeNode) int {
	if node == nil {
		return 0
	}
	left := postorder(node.Left)
	right := postorder(node.Right)
	tilt += Abs(right - left)
	return left + right + node.Val
}

func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: java/binary-tree-tilt.java
// 563. 二叉树的坡度 - Easy
// https://leetcode-cn.com/problems/binary-tree-tilt/
package java;


// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}


class BinaryTreeTilt {
    int tilt = 0;
    public int findTilt(TreeNode root) {
        postorder(root);
        return tilt;
    }

    public int postorder(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = postorder(node.left);
        int right = postorder(node.right);
        tilt += Math.abs(left - right);
        return left + right + node.val;
    }
}
